### ISSUE TITLE
Version catalog is an implicit input for all custom (as in not shipped with Gradle distribution) tasks types

### ISSUE THREAD

**Issue by erdi** [gradle/gradle#26020]
### Current Behavior

Any change to the version catalog, even one which does not affect any of the explicit inputs, renders all custom (as in not shipped with Gradle distribution) task types out of date.

### Expected Behavior

Only changes to the version catalog which affect explicit inputs render a task out of date.

### Context (optional)

This is especially painful for large builds which use large version catalogs. In our case we have 1100+ projects and 400 libraries defined in the catalog. The catalog is updated frequently because we have a policy and automation to upgrade to the latest version of third party OSS libraries used in the build as soon after they are released as possible. Our build logic is located in an included build and written in kotlin therefore it is compiled using a custom task type, `org.jetbrains.kotlin.gradle.tasks.KotlinCompile`. This means that any time anything in the catalog is changed, even an entry which is only used in a single project, no tasks are up to date any longer and a full rebuild happens even though actual inputs have not changed for most tasks.

I'm sure this behaviour is not by design because it does not affect built-in task types, like for example `org.gradle.api.tasks.compile.JavaCompile`. That's because these task types are loaded by the classloader which loads Gradle classes. That classloader cannot load classes generated into `.gradle/<version>/dependencies-accessors` as they do not form part of the classpath for it and the classloader classpath is in turn hashed and used as a task input (by `org.gradle.internal.snapshot.impl.ClassImplementationSnapshot#appendToHasher` via `org.gradle.internal.execution.caching.impl.DefaultCachingStateFactory#createCachingState`). This is in contrast to custom task types which are loaded by the build script classloader which has the output of `.gradle/<version>/dependencies-accessors` as part of its classpath and therefore, unintentionally, inputs of such tasks change every time the version catalog changes.

### Steps to Reproduce

1. Create a sample project build and apply kotlin jvm plugin to it.
2. Add a dummy kotlin class to the main source set.
3. Add a toml version catalog under `gradle/libs.versions.toml` with an entry for a library which does not need to be referenced anywhere in the build.
4. Run the `compileKotlin` task.
5. Rename the entry in the catalog.
6. Run the `compileKotlin` task again. It should be up to date but isn't.

If you follow the above steps but use `java` plugin, a java class and `compileJava` task instead then that task will be up to date even after renaming the entry in the catalog.

### Gradle version

8.2.1

**Comment by melix** [comment]
This isn't a bug but the consequence of the design. Version catalogs are an implicit input, because they generate classes which are on the build script classpath, so that users can have type-safe accessors. If the catalog changes, the classpath changes, so the scripts _may_ or _may not_ compile, we don't know. I would say it's more a problem of incremental compilation of the build scripts than it is of catalogs.

Note that this was known since day 1 when designing catalogs. They were designed to replace one particular popular pattern in the Android ecosystem, which was to declare versions in `buildSrc`. But the situation was much worse there, since _any_ change to the catalog would trigger recompilation. With version catalogs, it's only when a dependency is added, or that coordinates are changed (but NOT versions) that compilation would be invalidated. That was done with the assumption that you change dependency versions much more often than you change dependencies themselves.

One solution to mitigate this is to split very large multiprojects into composite builds. This will bring several advantages, but also drawbacks. There's no silver bullet.

**Comment by erdi** [comment]
Thanks for your response, @melix, it's highly appreciated. Also thanks for the insight about what pattern the version catalogs were designed to replace, I did not know that.

So there are two concerns here. One, which you are talking about, is that the build scripts need to recompile every time a dependency is added, removed, its coords are changed or its renamed in the catalog. I understand why that is, it's a fair requirement and I would expect that will be the case because I know of the generated accessors. All good. But there is another concern here and it's something that the issue I opened is about - a change to the catalog will also render any task type which is not part of Gradle distribution out of date even if no inputs for that task have actually changed. That's because the "implementation" input that Gradle implicitly adds contains a hash of the classpath for the classloader that loaded the task type as I, admittedly in a convoluted way, explained in the issue description. And in case of tasks not shipped with Gradle that class loader is the build script class loader which has the generated accessors on its classpath. This is what feels like big limitation if not a bug and is extremely painful for very large builds.

To sum up: I understand why build scripts need to recompile after a catalog change because they might access the generated accessors but a third party task type should not need to rerun because there is no way its implementation should or even can access the generated accessors for the catalog.

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

***

Related:
* https://github.com/gradle/gradle/issues/25616
* https://github.com/gradle/gradle/issues/6045

**Comment by antohaby** [comment]
As for workaround try adding KGP to buldscript classpath directly.

```
// settings.gradle.kts
buildscript {
    repositories { mavenLocal(); mavenCentral(); }
    dependencies {
        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:<YOUR_KOTLIN_VERSION>")
    }
}
```

This way KGP classes will have stable classloader that doesn't depend on libs.versions.toml and therefore kotlin compile tasks should not get out of date on libs.versions.toml change.

**Comment by Vampire** [comment]
Important for the work-around of @antohaby is, that it has to be in the settings script.

In the build script it will not work, as the version catalog accessors are added to the `buildSrc` class loader which is a parent of all the build script class paths and thus changing anything that influences the generated accessors (everything except versions) has the same effect as changing anything in `buildSrc`, making all 3rd party tasks out-of-date and a cache-miss as their runtime classpath changed.

Alternatively to the `buildscript` block in the settings script, you can of course simply us the `plugins { ... }` block of the settings script with `apply false`. That it is a project plugin doesn't matter as this only adds it to the classpath but does not apply it.

**Comment by big-guy** [comment]
We've talked about changing the way third party plugins are loaded so they would be isolated from these kinds of changes.

We're not planning to address this for 8.x, unfortunately.

**Comment by inorichi** [comment]
Any update on this? We're already on 9.x and this is still an issue.

The workaround of adding the plugins in settings.gradle with `apply false` has the drawback of invalidating everything whenever any version of those plugins change, so you have to be mindful of the plugins you add there, as some may be used in a single module and you don't want to invalidate everything when that plugin is updated.

Compared to not using libs.versions where everything works as you would expect, it's a bit annoying when this happens on big projects 🫤
